frontend/src/components/formula/formula.jsx

```diff
@@ -1,55 +1,64 @@
 import { useState, useEffect } from 'react';
+import { store } from 'store';
+import { useContext } from 'use-context-selector';
 
-import { Node } from 'components';
-import { NODE } from 'utils/constants';
-
-const Formula = ({ args, setOutput, boxId }) => {
+const Formula = ({ args, setOutput, boxId, inputSetter = () => {} }) => {
     const identityFn = '(x) => x;';
-    const [formula, setFormula] = useState(identityFn);
+    const [formulaText, setFormulaText] = useState(identityFn);
     const [editable, setEditable] = useState(false);
+    const { dispatch } = useContext(store);
+
     const enable = () => setEditable(true);
     const disable = () => setEditable(false);
+    const saveFormula = (formula) => dispatch(({
+        type: 'SET_BOX_FORMULA',
+        boxId,
+        formula,
+    }));
+
+    const createFunction = () => (
+        // eslint-disable-next-line no-new-func
+        new Function(`return ${formulaText}`)
+    );
+
     const save = () => {
-        const formulaInput = document.getElementById(`${boxId}-input`);
-        setFormula(formulaInput.value);
+        const func = createFunction();
+        saveFormula(func);
         disable();
     };
+
     useEffect(() => {
+        setFormulaText(identityFn);
         if (!args) return null;
-        const func = `return ${formula}`;
-        // eslint-disable-next-line no-new-func
-        const result = new Function(func);
-        setOutput(result()(args));
-    }, []);
-        
+        const func = createFunction();
+        const output = func()(args);
+        setOutput(output);
+        console.log('inputSetter', inputSetter);
+        inputSetter(output);
+    }, [args]);
+
     const handleChange = (e) => {
-        setFormula(e.target.value);
-    }
+        setFormulaText(e.target.value);
+    };
 
     const calculate = () => {
         if (!args) return null;

```

Comment:
This function does some things in common with the useEffect that's watching args - the commonalities should be moved to a convenience function to prevent repetition.